atc/exec/set_pipeline_step.go

```diff
@@ -0,0 +1,244 @@
+package exec
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/concourse/baggageclaim"
+	"io/ioutil"
+	"sigs.k8s.io/yaml"
+	"strings"
+
+	"code.cloudfoundry.org/lager"
+	"code.cloudfoundry.org/lager/lagerctx"
+	"github.com/concourse/concourse/atc"
+	"github.com/concourse/concourse/atc/creds"
+	"github.com/concourse/concourse/atc/db"
+	"github.com/concourse/concourse/atc/db/lock"
+	"github.com/concourse/concourse/atc/exec/artifact"
+	"github.com/concourse/concourse/vars"
+)
+
+//go:generate counterfeiter . SetPipelineDelegate
+
+type SetPipelineDelegate interface {
+	BuildStepDelegate
+
+	Initializing(lager.Logger)
+	Starting(lager.Logger)
+	Finished(lager.Logger, ExitStatus, db.ConfigVersion, db.ConfigVersion)
+
+	FindTeam() (db.Team, bool, error)
+}
+
+// SetPipelineStep sets a pipeline to current team. If pipeline_name specified
+// is "self", then it will self set the current pipeline. This step takes pipeline
+// configure file and var files from some resource in the pipeline, like git.
+type SetPipelineStep struct {
+	planID      atc.PlanID
+	plan        atc.SetPipelinePlan
+	metadata    StepMetadata
+	delegate    SetPipelineDelegate
+	lockFactory lock.LockFactory
+	succeeded   bool
+}
+
+func NewSetPipelineStep(
+	planID atc.PlanID,
+	plan atc.SetPipelinePlan,
+	metadata StepMetadata,
+	delegate SetPipelineDelegate,
+	lockFactory lock.LockFactory,
+) Step {
+	return &SetPipelineStep{
+		planID:   planID,
+		plan:     plan,
+		metadata: metadata,
+		delegate:    delegate,
+		lockFactory: lockFactory,
+	}
+}
+
+func (step *SetPipelineStep) Run(ctx context.Context, state RunState) error {
+	logger := lagerctx.FromContext(ctx)
+	logger = logger.Session("set-pipeline-step", lager.Data{
+		"step-name": step.plan.Name,
+		"job-id":    step.metadata.JobID,
+	})
+
+	step.delegate.Initializing(logger)
+
+	variables := step.delegate.Variables()
+	params, err := creds.NewParams(variables, step.plan.Params).Evaluate()
+	if err != nil {
+		return err
+	}
+
+	spParams, err := atc.NewSetPipelineParams(params)
+	if err != nil {
+		return err
+	}
+
+	repository := state.Artifacts()
+	artifactSource := setPipelineArtifactSource{
+		ctx:    ctx,
+		logger: logger,
+		params: spParams,
+		repo:   repository,
+	}
+
+	atcConfig, err := artifactSource.FetchConfig()
+	if err != nil {
+		return err
+	}
+
+	step.delegate.Starting(logger)
+
+	stdout := step.delegate.Stdout()
+
+	if spParams.PipelineName == "self" {
+		spParams.PipelineName = step.metadata.PipelineName
```

Comment:
I wonder if this syntax could be resolved at a higher level? :thinking: It feels weird to be peering into the metadata for this. For example, a one-off build could technically be constructed with a `set_pipeline` step, and one-off builds aren't associated to any pipeline, so this would end up being an empty string.

Maybe whoever constructs the step should be responsible for resolving `self` to the pipeline name?

Reply:
@vito I now resolve `self` in the `EvaluatiePlan` method, please if you are ok with that.